File: reverseNodesInKGroups.py
# Time Complexity : O(N)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        i = 0
        dummy = ListNode()
        dummy.next = head
        begin = dummy
        while head != None:
            i += 1
            if i % k == 0:
                begin = self.reverse(begin, head.next)
                head = begin.next 
            else:
                head = head.next
        return dummy.next
    def reverse(self, begin, end):
        prev = begin
        curr = begin.next 
        fast = curr.next 
        first = begin.next 
        while fast != end:
            curr.next = prev 
            prev = curr
            curr = fast
            fast = fast.next 
            
        curr.next = prev
        first.next = fast
        begin.next = curr
        return first

File: reverseNodesInKGroups.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach
// We reverse k elements in one batch, and move on to the next batch
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || k==0) return head;

        ListNode dummy = new ListNode();
        dummy.next = head;

        ListNode begin = dummy;
        ListNode end = head;

        int i=0;

        while (head != null) {
            i++;
            if (i%k == 0) {
                begin = reverse(begin, head.next);
                head = begin.next;
            }
            else {
                head = head.next;
            }
        }

        return dummy.next;
    }

    private ListNode reverse(ListNode begin, ListNode end) {
        ListNode prev = begin;
        ListNode first = begin.next;
        ListNode curr = first;
        ListNode fast = curr.next;

        while (fast != end) {
            curr.next = prev;
            prev = curr;
            curr = fast;
            fast = fast.next;
        }
        curr.next = prev;

        // change the pointers of the begin and end
        begin.next = curr;
        first.next = end;

        // return the first since we want to mark it as the new beginning
        return first;
    }
}